File: apps/server/agentcore/api/schemas/llm_window.py
"""Diagnostic LLM window projection schemas (§8.3 window_from_journal wire shape)."""

from typing import Literal

from pydantic import BaseModel, Field


class LlmWindowToolCallFunction(BaseModel):
    name: str
    arguments: str


class LlmWindowToolCall(BaseModel):
    id: str
    type: Literal["function"] = "function"
    function: LlmWindowToolCallFunction


class LlmWindowMessageLine(BaseModel):
    role: Literal["system", "user", "assistant", "tool"]
    content: str | None = None
    tool_calls: list[LlmWindowToolCall] | None = None
    tool_call_id: str | None = None
    reasoning_content: str | None = None
    # Diagnostic merge tag: ``context_blocks`` = opening user was rendered from the
    # structured ``run_context`` ContextBlock list (UI substitutes those segments and
    # offers「查看原始拼接」for this full ``content``). Absent / null for other messages.
    origin: str | None = None


class RunLlmWindowResponse(BaseModel):
    """One run's folded LLM input window for a turn (diagnostic replay)."""

    run_id: str
    available: bool = Field(
        description="False when the journal lacks execution facts to fold a window."
    )
    messages: list[LlmWindowMessageLine] = Field(default_factory=list)

File: packages/contract-rest-types/src/llm-window.ts
/**
 * Diagnostic LLM window REST contract.
 *
 * Aligns with `apps/server/agentcore/api/schemas/llm_window.py` and
 * `GET /v1/conversations/{id}/messages/{mid}/runs/{run_id}/llm-window`.
 */

export type LlmWindowRole = "system" | "user" | "assistant" | "tool";

export interface LlmWindowToolCallFunction {
  name: string;
  arguments: string;
}

export interface LlmWindowToolCall {
  id: string;
  type: "function";
  function: LlmWindowToolCallFunction;
}

export interface LlmWindowMessage {
  role: LlmWindowRole;
  content?: string | null;
  tool_calls?: LlmWindowToolCall[] | null;
  tool_call_id?: string | null;
  reasoning_content?: string | null;
  /**
   * Diagnostic merge tag from `run_head.user_origin`.
   * `context_blocks` = opening user was rendered from the structured `run_context`
   * ContextBlock list (UI substitutes those segments; full `content` is the
   * verbatim concat for「查看原始拼接」).
   */
  origin?: string | null;
}

/** `GET .../runs/{run_id}/llm-window` response. */
export interface RunLlmWindowResponse {
  run_id: string;
  available: boolean;
  messages: LlmWindowMessage[];
}
